Public/Sdk/Public/Transformers/Transformer.ReadPipGraphFragment.dsc

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Transformer {
+
+    /**
+     * Read in a fragment of the graph as a binary file and all all pip from it to the pip graph.
+     * name: Name of this pip graph fragment, used to return a handle to it.
+     * file: Binary file to read.
+     * dependencyNames: All other pip graph fragments this one depends on.
+     * Returns a handle to the fragment just created, which can be used to specify dependencies on other graphs.
+     */
+    @@public
+    export function readPipGraphFragment(name: string, file: SourceFile, dependencyNames: string[]): string {

```

Comment:
Maybe the type of `dependencyNames` (as well as the return type) should be some opaque type instead of string, to enforce people to call `readPipGraphFragment` in the correct order.

Otherwise, someone might be tempted to write something like this
```typescript
   export const frag1 = readPipGraphFragment("f1", f`graphs/f1.bin`, []);
   export const frag2 = readPipGraphFragment("f2", f`graphs/f2.bin`, [ "f1" ]);
```
which doesn't guarantee that `frag1` is loaded before `frag2`.

Reply:
Is there an example of how to write a simple type?  I just did string for ease, but would like to make it strongly typed.

---
In reply to: [289547224](https://github.com/microsoft/BuildXL/pull/394#discussion_r289547224) [](ancestors = 289547224)

Reply:
You check out the `[ServiceId](https://github.com/microsoft/BuildXL/blob/aae4be7f5600e4da9587515adb44fd79ec981fc2/Public/Sdk/Public/Transformers/Transformer.Service.dsc#L56-L57)` opaque type.